nums/core/linalg.py

```diff
@@ -256,6 +256,147 @@ def inv(app: ArrayApplication, X: BlockArray):
     return result
 
 
+def inv_uppertri(app: ArrayApplication, X: BlockArray):
+    # Inversion of an Upper Triangular Matrix
+    # Use the method described in https://www.cs.utexas.edu/users/flame/pubs/siam_spd.pdf
+    assert len(X.shape) == 2
+    assert X.shape[0] == X.shape[1], "This function only accepts square matrices"
+    single_block = X.shape[0] == X.block_shape[0] and X.shape[1] == X.block_shape[1]
+    nonsquare_block = X.block_shape[0] != X.block_shape[1]
+
+    # If X is single block or block size is non-square, then reshape BS to row_size // 4
+    if single_block or nonsquare_block:
+        X = X.reshape(block_shape=(X.shape[0] // 4, X.shape[0] // 4))
+
+    # Setup metadata
+    full_shape = X.shape
+    grid_shape = X.grid.grid_shape
+    block_shape = X.block_shape
+
+    R = X.copy()
+    Zs = app.zeros(full_shape, block_shape, X.dtype)
+
+    # Calculate R_11^-1
+    r11_oid = R.blocks[(0, 0)].oid
+    r11_inv_oid = app.cm.inv(
+        r11_oid, syskwargs={"grid_entry": (0, 0), "grid_shape": grid_shape}
+    )
+    R.blocks[(0, 0)].oid = r11_inv_oid
+    R_tl_shape = block_shape
+
+    # Continue while R_tl.shape != R.shape
+    while R_tl_shape[0] != full_shape[0] and R_tl_shape[1] != full_shape[1]:
+        # Calculate R11
+        R11_block = (
+            int(np.ceil(R_tl_shape[0] // block_shape[0])),
```

Comment:
You're using integer division here, so I think the `int(np.ceil(...` portion is redundant.